New timer design.

sofar opened the pull request:
I could honestly not make much sense of the timer implementation
that was here. Instead I've implemented the type of timer algorithm
that I've used before, and tested it instead.

This implementation uses dtime as passed to this function and
assumes it can be used for rough timekeeping, while giving sub
second precision. Since no os.time() is used, we have excellent
sub-second scheduling precision, but it should be assumed that
server ticks can cause ticks to be delayed for well over seconds
in case of server load.

Timers are guaranteed to execute, even if delayed well beyond their
scheduled time. Timers will never run before their allotted time
occurs.

If server load is high, all expired timers will trigger and
server lag may be increased. It is therefore important to
schedule smaller jobs more often, and use fractional numbers
and prime numbers to prevent coalescing of events.

sofar committed:
New timer design.

I could honestly not make much sense of the timer implementation
that was here. Instead I've implemented the type of timer algorithm
that I've used before, and tested it instead.

The concept is extremely simple: all timers are put in an ordered
list. We check every server tick if any of the timers have
elapsed, and execute the function associated with this timer.

We know that many timers by themselves cause new timers to be
added to this list, so we iterate *backwards* over the timer
list. This means that new timers being added while timers are
being executed, can never be executed in the same function pass,
as they are always appended to the table *after* the end of
the table, which we will never reach in the current pass over
all the table elements.

We switch time keeping to minetest.get_us_time(). dtime is
likely unreliable and we have our own high-res timer that we
can fix if it is indeed broken. This removes the need to do
any sort of time keeping.

tenplus1 commented:
Nicely done.

sofar commented:
This needs to be tested on windows. I'm pretty sure ```os.time()``` on android is safe, as it's Linux, but I don't know about windows. Any testers?

sofar commented:
I just confirmed in the documentation that on windows this returns the epoch() seconds, so I believe that testing would be nice to have done, but not necessarily required.

red-001 commented:
@sofar I use windows. I will test it soon.

sofar commented:
@HybridDog I believe you wrote the original code. Care to review?

kaeza commented:
Hmm, looks better than the original implementation, but still ways off.

Test mod:
```lua
minetest.register_on_joinplayer(function(player)
	player = player:get_player_name()
	local start = os.clock()
	minetest.after(5, function()
		local now = os.clock()
		print("after 5:", now-start)
		local start = os.clock()
		minetest.after(.3, function()
			local now = os.clock()
			print("after .3:", now-start)
		end)
	end)
end)
```
Original code:
```
after 5:        3.898 <-- !?
after .3:       5.431 <-- !?!?!?
```
Your patch:
```
after 5:        4.59  <-- Quite accurate
after .3:       1.024 <-- Not very accurate, but way better.
```

>The downside of this design is that all timers will fire at clock like precision on the edge of each second. This is an acceptable approach. With a better precision time source, we could introduce jitter.

[`os.clock`](http://www.lua.org/manual/5.1/manual.html#pdf-os.clock).

sofar commented:
@kaeza reading the documentation, I did not think that ```minetest.after``` ever had sub-1 second precision or that passing a non-integer was even "defined behavior".

It also is pretty clear that the original implementation had no chance to even get close to approaching usefulness on sub-second resolution, hence, I opted to not implement a timer that has that resolution.

However, I intend to add a new API for sub-second resolution, one that looks as follows:

```core.after_ticks(ticks, function, { ...}```

I have a functional version already (just in mod form) that will guarantee that the function gets run after ```ticks``` server ticks. This should be ideal for mod writers over a higher-resolution ```minetest.after``` function, since they generally only want to use this for really short lived events that are occasional, and furthermore, they generally want to honor that if the server admin changes their tick interval to be faster/slower, that their timers adjust as well.

Having both these functions side-by-side would provide enough resolution range to suit most needs.

kaeza commented:
@sofar I've checked, and indeed `lua_api.txt` [does not define](https://github.com/minetest/minetest/blob/882a89d65aa78b89cfaf2af054d15cc4c94ad022/doc/lua_api.txt#L2244) if fractional times are allowed (probably due to the terse nature of the document; it's more of a "cheat sheet" of sorts than proper documentation). However, the [first example](http://dev.minetest.net/minetest.after#Examples) on the Development Wiki explicitly uses fractional time.

I know the dev wiki is community-driven, so it may or may not be correct. Would like official input on the interface.

sofar commented:
@kaeza which is why I've asked @HybridDog to review and comment, as I think he imlpemented the original code.

Unless we actually get a higher res clock time (ms interval would be great, already), I don't think this can be pulled off, and we should consider the dual approach I suggested by adding a per-tick timer interface.

kaeza commented:
FWIW, `boats` mod in `minetest_game` seems to depend on fractional times on [two](https://github.com/minetest/minetest_game/blob/a88576afd1a27faf1a8036e2ed4c4b1e5c78ec90/mods/boats/init.lua#L61) [occasions](github.com/minetest/minetest_game/blob/a88576afd1a27faf1a8036e2ed4c4b1e5c78ec90/mods/boats/init.lua#L77), and I'm pretty sure many mods also depend on it.

IMHO, `.after` should support fractional times since it costs nothing, and doesn't break mods which were not broken already.

Also, HybridDog is not the original author of the code. He enhanced (or tried to, no offense) the already existing implementation. (I couldn't find its original author; my git-fu is sorely lacking in this topic).

sofar commented:
@kaeza Actually, I think I can fix this.

sofar commented:
Needs a bit of a rewrite...

sofar commented:
@kaeza please run your testcase on this, and see what values you get.

sofar commented:
@kaeza you can't use ```os.clock()``` to look at real time, it reports cpu time, not clock time.

cpu time is the time the processor has been running. So if it was sleeping, ```os.clock()``` doesn't increase.

replace it with ```os.time()``` in your test case. No, you can't get sub-second precision.

sofar commented:
Better test code:

```
minetest.register_on_joinplayer(function(player)
       local start = os.time()
       for i = 1, 50 do
               minetest.after(i / 10, function(i, s)
                       print(i/10 .. ":" .. os.time() - s)
               end, i, start)
       end
end)
```

HybridDog commented:
l think fractional time has always been possible.

l edited the minetest.after function that the globalstep doesn't constantly loop through all current jobs and tests each one if it expired, l did a typo and fixed it, which doesn't get merged https://github.com/minetest/minetest/pull/3163/files

A shorter code doesn't mean it works faster.

l think for testing the speed of the different minetest.afters, you need to decrease dtime, which makes the globalstep function gets executed more often and minetest.after works more accurate. dtime called length of a server tick: https://github.com/minetest/minetest/blob/master/minetest.conf.example#L977-L979

sofar commented:
@HybridDog 

The latest commit here implements fractional time.

```A shorter code doesn't mean it works faster```

no, but it means that it's highly likely less complex, less code maintenance and less likely to contain errors. 

As I posted in the other PR, I extensively profiled both algorithms and they perform exactly the same. The remaining difference is complexity.

kaeza commented:
@sofar With latest patch:

```
after 5:        3.224 <-- Quite bad precision
after .3:       0.257 <-- Excellent
```

I won't use `os.time` since the point is in supporting subsecond resolution which `os.time` cannot detect.

Can you try using [`minetest.get_us_time`](https://github.com/minetest/minetest/blob/master/doc/lua_api.txt#L1729)? It advertises microsecond resolution (uses `gettimeofday` on POSIX and `QueryPerformanceCounter` on Windows).

sofar commented:
@kaeza as I said before, using os.clock() is literally bullshit, you're not measuring what you think you are measuring.

Your measurements are therefore invalid.

sofar commented:
@kaeza I'll look into get_us_time(), hadn't seen it before until now

sofar commented:
@kaeza hhmmm, why don't you use minetest.get_us_time() then in your test mod code?

```
minetest.after(1, function(player)
       local start = minetest.get_us_time()
       minetest.after(5, function()
               local now = minetest.get_us_time()
               print("after 5:", (now-start)/1000000)
               local start = minetest.get_us_time()
               minetest.after(.3, function()
                       local now = minetest.get_us_time()
                       print("after .3:", (now-start)/1000000)
               end)
       end)
end)

```

outputs for me:

```
after 5:	5.010517
after .3:	0.301732
```

kaeza commented:
@sofar Here's my output under Windows 7 using `get_us_time`:
```
after 5:        3.799379
after 5:        3
after .3:       0.331271
```
(the second "after 5" is using os.time)

It's still a second too early for some reason. I give up for now.

sofar commented:
@kaeza ouch, wow, that's bad.

Note however that the timer itself neither uses os.* nor get_us_time()....

HybridDog commented:
Do you use any mods when you do the tests?

sofar commented:
@HybridDog yes, I have been using:

- lightning mod (has a long-range timer)
- the chest mod I wrote that uses a very short-time timer (0.1-0.2sec)
- tnt is builtin, so is leafdecay, both use timers

the whole reason I wrote the code is because I noticed that the original timer code was completely unreliable ;)

ShadowNinja commented on builtin/game/misc.lua:
```diff
@@ -4,74 +4,34 @@
+	table.insert(_jobs, {
+		func = func,
+		expire = _time + time,
+		arg = { ... } or {},
```
`{...} or {}` will never evaluate the second part.  `{...}` with no arguments is just `{}`.

ShadowNinja commented:
Note: any time sources used should be monotonic.  This means either `dtime` (although that seems to be inacurate and may not be monotonic either) or an API extension I think.

sofar committed:
Clocksource: use a better clock if available.

clock_gettime() is a far better clock than gettimeofday().

Even better than clock_gettime() is that you can select either
CLOCK_MONOTONIC, or even CLOCK_MONOTONIC_RAW. These clocks offer
high precision time. And the _RAW variant will never roll back
due to NTP drift or daylight savings, or otherwise.

I've adjusted this code to select the right clock method auto-
matically based on what's available in the OS. This means that
if you're running a very old linux version, MacOS or other,
you will automatically get the best clocksource available.

I've tested all Linux clocksources by selectively compiling and
running a 10k+ timer test suite. In all cases I confirmed that
the 3 POSIX Linux clocksources worked properly, and were
selected properly.

I've modified the OS X compile path to use the high-res clock
source for all time functions, but I can't confirm it works or
that it compiles.

As for WIN32, I confirmed that the used clocksource is indeed
a Monotonic clocksource, so good news: that code section appears
to be exactly what it should be.

sofar commented:
@ShadowNinja @kaeza Good news

I checked out porting.h and confirmed:

- WIN32 is using a monotonic clocksource properly
- OS X is appearing to do so as well

However, it was clear that POSIX using ```gettimeofday()``` is not MONOTONIC.

But, that is easily fixed. And so there are two patches in this PR now:

1. adjusts ```minetest.after()``` to internally use ```core.get_us_time()```

and snice ```core.get_us_time()``` uses the code from ```src/porting.h```

2. a patch that uses ```clock_gettime()``` with ```CLOCK_MONOTONIC_RAW``` or (if not available) ```CLOCK_MONOTONIC```, with a fallback to ```gettimeofday```

I confirmed on Linux that all 3 those clocksources correctly work and pass my timer tests. Since I patched *all* the time source code in the game, *all* time derivative code should improve with this change.

This needs *testing on OS X* as I can't test it. The code change for OS X is trivial, but it still needs testing.

sofar commented:
Don't merge this yet. I need to account for clock rollover still.

ShadowNinja commented:
Shouldn't you just switch to a `u64`?
The docs should define `get_us_time` as being monotonic.
`get_us_time` isn't really a good name for a monotonic time source though -- system clock time is implied `get_steady_time`?
I also think that simply `get(_steady)_time` would be a better API -- it would return the time in seconds as a floating point number.  This would allow for arbritrary precision, and is how Python does it.

Also, this is a case where C++11's chrono functions whould be useful.  IIRC we decided to switch after Debian stable was updated, and then some people started saying it was after Ubuntu LTS got updated.

sofar commented:
@ShadowNinja switching to u64 doesn't necessarily accomplish much, except make rollover much more rare.

```get_us_time``` is not necessarily monotonic, as playing on a server that doesn't support ```CLOCK_MONOTONIC``` does not guarantee it being monotonic, so at best we can state that it may be monotonic in most OS's.

There is little need to add a new ```get_steady_time```. All the current users of get_us_time will see no difference, only an improvement. Having another interface for no users doesn't accomplish anything at this point.

There is not a single mod or piece of code that currently relies on the non-monotonicity of the clock.

I'm not saying this shouldn't be improved upon further, but rewriting it for the sake of rewriting doesn't accomplish much. This set of patches does a good job in improving the current situation drastically while maintaining compatibility with every OS that's supported now... I'd prefer that myself.

sofar commented:
Reversed order of the commits. Added rollover accounting (tested).

kaeza commented:
I changed a bit the test code:

```
local function test(callcount)
	local start = minetest.get_us_time()
	local start_t = os.time()
	minetest.after(5, function()
		local now = minetest.get_us_time()
		print("after 5:", (now-start)/1000000)
		print("after 5:", os.time()-start_t)
		local start = minetest.get_us_time()
		minetest.after(.3, function()
			local now = minetest.get_us_time()
			print("after .3:", (now-start)/1000000)
			if callcount<5 then
				return test(callcount+1)
			end
		end)
	end)
end

minetest.register_on_joinplayer(function(player)
	player = player:get_player_name()
	test(1)
end)
```

And tested this patch:

```
after 5:        4.303367
after 5:        5
after .3:       0.324845
after 5:        4.999138
after 5:        5
after .3:       0.325674
after 5:        5.007121
after 5:        5
after .3:       0.317796
after 5:        5.038544
after 5:        5
after .3:       0.316283
after 5:        4.9998
after 5:        5
after .3:       0.324163
```

Looks much better.

ShadowNinja commented:
@sofar: 735 thousand year rollover is more than just "much more rare", it's practically impossible.

It may mot always be monotonic, but it should be specified as not neccessarily being wall time, as I'd just expect it to be basically `time(nullptr)*1000` (asuming floating point time()) otherwise.

sofar commented:
I'm not going to spend time rewriting time to u64 until after this is merged. Rewriting internal time keeping to u64 everywhere is likely going to touch code all over the place, and I have not much confidence that this patch is getting any positive reaction so far, even though all I have done is exposed  bad code and provided improvements that can be built upon further. (And that's the way it should be done).

So, unless I get some confidence that this patch gets the necessary review, and a few core devs being appreciative of the effort, I'm just going to watch this one out.

```
grep -r getTime.s  .|wc -l
49
```

ShadowNinja commented on builtin/game/misc.lua:
```diff
@@ -4,74 +4,42 @@
+		_time = _time + (_new - _last)
+	else
+		-- rollover
+		_time = _time + _new
```
This isn't quite correct, the real dtime will be a bit more (`UINT32_MAX / 1000000 - last` more, if my math's right).
I suppose it doesn't matter much if this only happens rarely, but I'd fall back to `dtime` in this case (assuming `dtime` doesn't go haywire on overflow).

sofar replied:
Right, this would mean we'd lose up to a server tick in time, potentially. It's not critical by any means, and in fact means that all the timers slide forward in time (as in, they'd run slightly faster). Hence this seems acceptable - having timers run 0.1 seconds faster once in ~2 hours doesn't hurt anything significant.

ShadowNinja commented on builtin/game/misc.lua:
```diff
@@ -4,74 +4,42 @@
-end
+local _jobs = {}
+local _time = 0.0
+local _last = 0.0
```
What's with all the underscores?  These are all local, so they can't be seen by anything outside the file.

sofar replied:
habit, it makes spotting locals a bit more apparent. In various C styles, _ and __ are using to delimit "internal" stuff that's not supposed to leak or is available out of context. I can remove these, obviously.

ShadowNinja replied:
Please do.

ShadowNinja commented on builtin/game/misc.lua:
```diff
@@ -4,74 +4,42 @@
-	if delay < mintime then
-		return
+
+	for i = #_jobs, 1, -1 do
```
It seems like this is done in reverse order so that the `table.remove` call doesn't mess things up, but that's not immediately apparent.  Could you add a comment describing why it's in reverse order?

sofar replied:
Right, that's the crux of the rewrite right there, happy to comment that.

ShadowNinja commented on src/porting.h:
```diff
@@ -211,43 +211,58 @@ void initIrrlicht(irr::IrrlichtDevice * );
 
 #else // Posix
-
-	inline u32 getTimeS()
```
These should also be documented as not (necessarily) being wall time.

sofar replied:
I'll see in what places I can adjust the gap in documentation. This may just be a reason to get that wiki account.

ShadowNinja commented:
Few nits, otherwise looks great!  :+1:

kaeza commented on src/porting.h:
```diff
@@ -211,43 +211,58 @@ void initIrrlicht(irr::IrrlichtDevice * );
+		ts.tv_sec = mts.tv_sec;
+		ts.tv_nsec = mts.tv_nsec;
+#else
+#ifdef CLOCK_MONOTONIC_RAW
```
`#elif defined(...)`, so you don't have to stack `#endif`s at the end.

sofar replied:
sure, will do.

Jordach replied:
-----Original Message-----
From: "Auke Kok" <notifications@github.com>
Sent: ‎27/‎01/‎2016 16:52
To: "minetest/minetest" <minetest@noreply.github.com>
Subject: Re: [minetest] New timer design. (#3593)

In src/porting.h:
>  	{
> +#if defined(__MACH__) && defined(__APPLE__)
> +	// from http://stackoverflow.com/questions/5167269/clock-gettime-alternative-in-mac-os-x
> +	// OS X does not have clock_gettime, use clock_get_time
> +		clock_serv_t cclock;
> +		mach_timespec_t mts;
> +		host_get_clock_service(mach_host_self(), CALENDAR_CLOCK, &cclock);
> +		clock_get_time(cclock, &mts);
> +		mach_port_deallocate(mach_task_self(), cclock);
> +		ts.tv_sec = mts.tv_sec;
> +		ts.tv_nsec = mts.tv_nsec;
> +#else
> +#ifdef CLOCK_MONOTONIC_RAW
sure, will do.
—
Reply to this email directly or view it on GitHub.

sofar commented:
- Added documentation about ```minetest.after()``` allowing fractional time.
- Added documentation on ```get_us_time()``` not being wall clock time.
- Added a comment in the time code about not using wall time.
- Removed underscores
- use #if defined() end #elif.

paramat commented:
If ShadowNinja is happy i'll +1 and merge this, this will enable a nice behaviour for the new chests PR.

kilbith commented on builtin/game/misc.lua:
```diff
@@ -4,74 +4,45 @@
+		if time >= jobs[i].expire then
+			core.set_last_run_mod(jobs[i].mod_origin)
+			jobs[i].func(unpack(jobs[i].arg))
+			table.remove(jobs, i)
```
Why not using `jobs[i] = nil` ? This is a bit faster.

rubenwardy replied:
It doesn't remove the key or shift the keys after it down.

kilbith replied:
Wrong. It's doing the same job.

![capture d ecran de 2016-01-27 23-40-58](https://cloud.githubusercontent.com/assets/7883281/12631042/a15baaac-c54f-11e5-8556-385b4c358210.png)

sofar replied:
Right, we can't use that method here. While it is indeed faster, it would leave a 'nil' hole in the table.

kaeza replied:
@kilbith If `jobs[i].func` (i.e. the `after` callback) calls `minetest.after` itself, it would leave a `nil` "hole" in the `jobs` array breaking the whole thing.

(`table.remove` could be localized to save a few lookups, but will probably not yield a big enough performance boost anyway).

kaeza replied:
What could be done here is localize `jobs[i]` since it's looked up 4 times each iteration.

sofar replied:
Given that this runs 10-20 times per second, I'll patch that in.

sofar commented:
updated with @kaeza 's comment storing jobs[i] for reuse in hotpath.

kaeza commented:
:+1:

ShadowNinja commented:
4ac1e9bccbd0222ab779b3fc4d2144d60e1f2b49, ad884f23d4e73d38ce0f529de49591dd66cee44d

neoascetic commented:
Please fix OSX build, here the logs: https://travis-ci.org/krondor-game/minetest/builds/105610917

neoascetic commented:
#3628 fixes the issue

est31 commented:
Mhh, perhaps having OSX in travis might be an idea?

neoascetic commented:
@est31 I am working on this, travis now supports multiple OSes. Please merge #3628 and I'll make a PR.
